Update docs for version 7.0.0, remove old APIs

mbland opened the pull request:
### Description

Describes the new `scala_toolchains()` API, breaking changes, Bazel compatibility matrix, and breaking of the `io_bazel_rules_scala` repo name dependency. Part of #1482, #1647, #1652, and #1699, it also removes obsolete `WORKSPACE` APIs in favor of `scala_toolchains()`, et. al.

Much of the README text regarding `WORKSPACE` configuration, Bazel compatibility, and breaking changes comes from:

- https://github.com/bazelbuild/rules_scala/issues/1482#issuecomment-2605683509

The "Breaking changes in `rules_scala` 7.x" section of `README.md` describes the removed `WORKSPACE` macros, and provides guidance for adopting the `scala_toolchains()` API. Also includes information about the upcoming Bzlmod implementation (#1482) and the Bazel 8 compatibility changes for `rules_scala` 8.x (#1652).

Contains a light editing pass over almost every Markdown file to resolve `markdownlint` warnings when editing in VSCode. Also added `.markdownlint.json` to silence rule `MD033/no-inline-html`, since several docs contain `<br/>` and/or `<table>` elements.

- https://github.com/DavidAnson/vscode-markdownlint?tab=readme-ov-file#markdownlintconfig

Performed other opportunistic grammar edits and added new information, including:

- `docs/coverage.md`: Describes how to determine the default JaCoCo version used by `rules_java`

- `scripts/README.md`: Updates `create_repository.py` docs extensively, and adds a section for `sync_bazelversion.sh`

- Replaces `starlark` code fences with `py`, since the syntax is identical and editors seem to support it better.

### Motivation

Since the next major release is imminent, ensuring the documentation accurately reflects all the changes since v6.6.0 has become an urgent priority.

Rather than leave the old `WORKSPACE` APIs in place, which could lead to surprising failures, this change removes all of them. This also changes some code that still depended on some of these obsolete macros, including `scala_toolchains()`. Since all the toolchainization changes led to the entire project already using `scala_toolchains()` and `scala_register_toolchains()` exclusively, this proved a low risk refactoring.

With some Bzlmod and Bazel 8 information already in place, very little will need to change when these implementations land. The commits that contain those implementations will also include their relevant documentation updates.

bartoszkosiorek commented on README.md:
```diff
@@ -15,101 +17,167 @@ This project defines core build rules for [Scala](https://www.scala-lang.org/) t
+[Install Bazel]: https://docs.bazel.build/versions/master/install.html
+[Bazelisk]: https://docs.bazel.build/versions/master/install.html
+
+Add the following stanza to your `WORKSPACE` file and update versions with their
```
Can you explain what is `stanza`?
Unfortunately it is not common word. Maybe you could replace it with `code` or `code snippet`?
I was a little bit confused about it at first :-D

mbland replied:
Ah, sorry. Updated in https://github.com/bazelbuild/rules_scala/pull/1703/commits/b294878b07c875f2e1ca374f11e12c88593bb74e.

gergelyfabian replied:
https://en.wikipedia.org/wiki/Stanza

gergelyfabian replied:
Seems to be used originally in the context of literature, but it is also commonly used in the context of IT for a group of lines (similarly as in literature).

bartoszkosiorek reviewed (approved):
Great job!
It is well written, and explain migration details.
Thanks.

mbland commented:
> Great job! It is well written, and explain `rules_scala 7.0` migration details. Thanks.

@bartoszkosiorek Thanks!

Also, I've added a section about having to use `use_repo()` to bring repos into scope under Bzlmod, based on the discussion with @michalbogacz mentioned in https://github.com/bazelbuild/rules_scala/pull/1703/commits/be73da75c5ab61f8db0602293deec5e5ab330162.

That discussion also got me to realize that the `native.toolchain()` call in `setup_scala_toolchain()` could likely move into the `scala_toolchain()` wrapper macro, to avoid the explicit `toolchain()` call mentioned in this latest update. I'll open a new pull request for that (if it works, which I expect it will).

gergelyfabian commented on docs/coverage.md:
```diff
@@ -55,26 +55,78 @@ echo "coverage report at file://${destdir}/index.html"
-The current Jacoco version in Bazel (0.8.3) has missing coverage for lambdas
-(including for comprehensions; see issue https://github.com/bazelbuild/rules_scala/issues/1056). Also, the support for
-filtering out code generated by the Scala compiler is quite reduced in Jacoco.
+`rules_scala` uses the [JaCoCo](https://www.jacoco.org/jacoco/) library imported
```
Thank you for adding this!

mbland replied:
My pleasure! Thanks for taking such a thorough look over everything!

gergelyfabian commented on docs/cross-compilation.md:
```diff
@@ -113,43 +131,49 @@ scala_library(
-
 ## Requesting specific version
+
 To use other than default version of Scala, you need to change the current `@io_bazel_rules_scala_config//:scala_version` build setting.
```
Probably this line also needs to change.

mbland replied:
Pulling this thread destroyed my sweater... I've overhauled the entire document in https://github.com/bazelbuild/rules_scala/pull/1703/commits/427e1c9f92e4ae8252214565c5542e1a21a849e9. Please take a look and tell me what you think. I think most of it is OK, but I didn't do much to the "Cross-build support tiers" section at the end.

In the process, I learned that ['incompatible_use_toolchain_transition' has been deprecated since before Bazel 6.5.0](https://bazel.build/versions/6.4.0/rules/lib/globals?hl=en#rule). I took the liberty to remove that attribute throughout the project, and made sure `./test_all.sh` passed before including it in the same commit.

gergelyfabian replied:
I looks great, thanks!

gergelyfabian commented on docs/cross-compilation.md:
```diff
@@ -113,43 +131,49 @@ scala_library(
 scala_library(
     ...
     srcs = select({
         "@io_bazel_rules_scala_config//:scala_version" + version_suffix(v): srcs(v)
```
I think there are still a lot of "io_bazel_rules_scala" instances laying around in the docs.

mbland replied:
Not really, though I did double check and found two more that I've since updated in https://github.com/bazelbuild/rules_scala/pull/1703/commits/5bb45294a140ae45c372d4ea620692b7f8a0225b (one in `README.md`, another in `.github/workflows/workspace_snippet.sh`).

As mentioned in that commit message, the remaining references are actually prefixed `io_bazel_rules_scala_` (notice the trailing `_`). Most of them are private, with the exception of `@io_bazel_rules_scala_config`, which is a documented public interface.

I also mention that we _could_ change this to `@rules_scala_config` as part of the breaking changes for v7.0.0. If @simuons and/or @liucijus say the word, I'm happy to do that in a separate pull request.

gergelyfabian replied:
Thanks, makes sense.

simuons replied:
@mbland I'm ok with renaming `@io_bazel_rules_scala_config` since there will be breaking changes as you mentioned and there is no reason for keeping label misaligned.

mbland replied:
@simuons Thanks, I'll take care of that in a follow-up PR. That'll be another sweeping, albeit straightforward change similar to #1696, and this current PR has already gotten bit enough.

mbland replied:
Preview: [rules-scala-config-repo-name in mbland/rules_scala](https://github.com/mbland/rules_scala/tree/rules-scala-config-repo-name) (specifically https://github.com/mbland/rules_scala/commit/d95c0e7ba6d1500df406e688e62d06c3229e85c0).

It wasn't _as_ big a diff as #1696, but still probably worth its own PR.

gergelyfabian commented on README.md:
```diff
@@ -15,101 +17,172 @@ This project defines core build rules for [Scala](https://www.scala-lang.org/) t
+
+rules_scala_dependencies()
+
+load("@rules_java//java:rules_java_deps.bzl", "rules_java_dependencies")
```
I tried doing this, and got:

```
ERROR: Error computing the main repository mapping: cannot load '@@rules_java//java:rules_java_deps.bzl': no such file
```

mbland replied:
D'oh! Thanks for that catch. Fixed in https://github.com/bazelbuild/rules_scala/pull/1703/commits/85519d06bbef243b8099ddf76ed6c75e96d8b5ff.

As I mention in that commit message, this was the correct snippet for `rules_java` 8.x, used in my Bazel 8 compatibility branch. It's now updated for `rules_java` 7.x, matching what's in the current `WORKSPACE` files, with some comments explaining what to expect when `rules_java` 8.x lands in this repo.

gergelyfabian replied:
Probably we'd need a documentation change when we'd land `rules_java` 8.x (maybe add that on a branch already?).

mbland replied:
Of course! I've already been updating the `README.md` in [bzlmod-bazel-8](https://github.com/mbland/rules_scala/tree/bzlmod-bazel-8). 😉 As I've described in several of my "weekly updates" from #1482, I've been keeping that branch and [bzlmod](https://github.com/mbland/rules_scala/tree/bzlmod) up to date with the latest merges to `master` and other developments.

This PR contains the bulk of the necessary doc updates, so it's a much lighter lift to include doc updates in those other branches now.

mbland commented on README.md:
```diff
@@ -15,101 +17,172 @@ This project defines core build rules for [Scala](https://www.scala-lang.org/) t
+    name = "rules_scala",  # Can be "io_bazel_rules_scala" if you still need it.
+    sha256 = "<SHASUM>",
+    strip_prefix = "rules_scala-7.0.0",
+    url = "https://github.com/bazelbuild/rules_scala/releases/download/7.0.0/rules_scala-7.0.0.tar.gz",
```
Good point. I've updated the `7.0.0` references to `<VERSION>` in https://github.com/bazelbuild/rules_scala/pull/1703/commits/5bb45294a140ae45c372d4ea620692b7f8a0225b. Let me know what you think of that solution.

simuons reviewed (approved):
Thanks, @mbland! Really appreciated. Well done indeed. Looks like there is a conflict, I'll merge once you will resolve it.

Special thanks @bartoszkosiorek and @gergelyfabian for looking into this.

mbland commented:
> Thanks, @mbland! Really appreciated. Well done indeed. Looks like there is a conflict, I'll merge once you will resolve it.
> 
> Special thanks @bartoszkosiorek and @gergelyfabian for looking into this.

My pleasure! I've rebased on `master` after #1700, and added https://github.com/bazelbuild/rules_scala/pull/1703/commits/c789cb024a78bb2655e3cbdb6b29b68b3f0ecc70 as one final change to address the decision from https://github.com/bazelbuild/rules_scala/issues/1482#issuecomment-2684792558 to remove the `testing` parameter from `scala_toolchains()`.
